nullaway/src/main/java/com/uber/nullaway/generics/PreservedAnnotationTreeVisitor.java

```diff
@@ -91,4 +94,99 @@ public Type visitParameterizedType(ParameterizedTypeTree tree, Void p) {
   protected Type defaultAction(Tree node, Void unused) {
     return castToNonNull(ASTHelpers.getType(node));
   }
+
+  private interface TypeMetadataBuilder {
+    TypeMetadata create(com.sun.tools.javac.util.List<Attribute.TypeCompound> attrs);
+
+    Type cloneTypeWithMetaData(Type typeToBeCloned, TypeMetadata metaData);
+  }
+
+  private static class JDK17AndEarlierTypeMetadataBuilder implements TypeMetadataBuilder {
+
+    @Override
+    public TypeMetadata create(com.sun.tools.javac.util.List<Attribute.TypeCompound> attrs) {
+      return new TypeMetadata(new TypeMetadata.Annotations(attrs));
+    }
+
+    @Override
+    public Type cloneTypeWithMetaData(Type typeToBeCloned, TypeMetadata metaData) {
+      return typeToBeCloned.cloneWithMetadata(metaData);
+    }
+  }
+
+  private static class JDK21TypeMetadataBuilder implements TypeMetadataBuilder {
+
+    private static final MethodHandle typeMetadataHandle = createHandle();
+    private static final MethodHandle cloneAddMetadataHandle =
+        createVirtualMethodHandle(Type.class, TypeMetadata.class, Type.class, "addMetadata");
+    private static final MethodHandle cloneDropMetadataHandle =
+        createVirtualMethodHandle(Type.class, Class.class, Type.class, "dropMetadata");
+
+    private static MethodHandle createHandle() {
+      MethodHandles.Lookup lookup = MethodHandles.lookup();
+      MethodType mt = MethodType.methodType(void.class, com.sun.tools.javac.util.ListBuffer.class);
+      try {
+        return lookup.findConstructor(TypeMetadata.Annotations.class, mt);
+      } catch (NoSuchMethodException e) {
+        throw new RuntimeException(e);
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private static MethodHandle createVirtualMethodHandle(
+        Class<?> retTypeClass, Class<?> paramTypeClass, Class<?> refClass, String methodName) {
+      MethodHandles.Lookup lookup = MethodHandles.lookup();
+      MethodType mt = MethodType.methodType(retTypeClass, paramTypeClass);
+      try {
+        return lookup.findVirtual(refClass, methodName, mt);
+      } catch (NoSuchMethodException e) {
+        throw new RuntimeException(e);
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public TypeMetadata create(com.sun.tools.javac.util.List<Attribute.TypeCompound> attrs) {
+      ListBuffer<Attribute.TypeCompound> b = new ListBuffer<>();
+      b.appendList(attrs);
+      try {
+        return (TypeMetadata) typeMetadataHandle.invoke(b);
+      } catch (Throwable e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public Type cloneTypeWithMetaData(Type typeToBeCloned, TypeMetadata metaData) {
+      try {
+        // In Jdk21 addMetadata works if there is no metadata associated with the type, so we create
+        // a copy without the existing metadata first and then add it
```

Comment:
Is it possible that deleting the extant metadata will cause some important info to be lost, unrelated to nullability?  Not sure how easily we can answer this, just curious

Reply:
The way I understand it from the documentation, it removes the Metadata associated with the TypeMetadata class that we pass as a parameter(and keeps the rest of the Metadata), and we are naturally passing in the class for which we want to add the Metadata. I am not entirely sure if there would be some important information that might be lost I might have to dig deeper for that.